File: Problem9.java
//Time complexity: O(logn)
//Space Complexity: O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

public class Problem9 {
    public int findPeakElement(int[] nums) {
        int low = 0;
        int high = nums.length-1;
        if(nums.length==1)
            return 0;
        while(low<=high){
            int mid = low + (high-low)/2;
            if(mid == 0){
                if(nums[mid]>nums[mid+1])
                    return mid;
                else
                    return mid+1;
            }
            if(mid == nums.length-1){
                if(nums[mid]>nums[mid-1])
                    return mid;
                else
                    return mid-1;
            }

            if(nums[mid]>nums[mid-1] && nums[mid]>nums[mid+1])
                return mid;
            if(nums[mid+1]>nums[mid])
                low = mid+1;
            else
                high = mid;
        }
        return -1;
    }
}

File: Problem_9.py
# Time Complexity: O(log n) + O(log n) = 2*O(log n) ~ O(log n)
# Space Complexity: O(1)

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        left, right = 0, len(nums) - 1
        idx1 = idx2 = - 1

        if nums == []:
            return [idx1, idx2]
            
        if target == nums[len(nums)-1] and target == nums[0]:
            return [0, len(nums)-1]

        while left <= right:
            mid = left + (right - left // 2 )
            if target > nums[mid]:
                left = mid + 1
            elif target < nums[mid]:
                right = mid - 1
            else:
                idx1 = mid
                right = mid - 1

        left, right = 0, len(nums) - 1
        while left <= right:
            mid = left + (right - left // 2 )
            if target > nums[mid]:
                left = mid + 1
            elif target < nums[mid]:
                right = mid - 1
            else:
                idx2 = mid
                left = mid + 1
        
        return [idx1, idx2]